### PR TITLE
Fixes #432 ...

### LINKED ISSUE #432: Segfault interpreting a function with ccall the second time
This is the complete session:
```
julia> using JuliaInterpreter

julia> function f()
           t = @ccall time()::Cint
       end
f (generic function with 1 method)

julia> @interpret f()
1600247465

julia> @interpret f()

signal (11): Segmentation fault
in expression starting at REPL[2]:1
do_apply at /home/fredrik/julia15/src/builtins.c:523
bypass_builtins at /home/fredrik/.julia/packages/JuliaInterpreter/muncv/src/interpret.jl:179
#evaluate_call_recurse!#58 at /home/fredrik/.julia/packages/JuliaInterpreter/muncv/src/interpret.jl:203
evaluate_call_recurse! at /home/fredrik/.julia/packages/JuliaInterpreter/muncv/src/interpret.jl:202 [inlined]
eval_rhs at /home/fredrik/.julia/packages/JuliaInterpreter/muncv/src/interpret.jl:389
step_expr! at /home/fredrik/.julia/packages/JuliaInterpreter/muncv/src/interpret.jl:532
step_expr! at /home/fredrik/.julia/packages/JuliaInterpreter/muncv/src/interpret.jl:582
unknown function (ip: 0x7f9521541a91)
finish! at /home/fredrik/.julia/packages/JuliaInterpreter/muncv/src/commands.jl:14
finish_and_return! at /home/fredrik/.julia/packages/JuliaInterpreter/muncv/src/commands.jl:29 [inlined]
finish_and_return! at /home/fredrik/.julia/packages/JuliaInterpreter/muncv/src/commands.jl:33 [inlined]
finish_and_return! at /home/fredrik/.julia/packages/JuliaInterpreter/muncv/src/commands.jl:33
_jl_invoke at /home/fredrik/julia15/src/gf.c:2214 [inlined]
jl_apply_generic at /home/fredrik/julia15/src/gf.c:2398
jl_apply at /home/fredrik/julia15/src/julia.h:1690 [inlined]
do_call at /home/fredrik/julia15/src/interpreter.c:117
eval_value at /home/fredrik/julia15/src/interpreter.c:206
eval_body at /home/fredrik/julia15/src/interpreter.c:428
jl_interpret_toplevel_thunk at /home/fredrik/julia15/src/interpreter.c:660
jl_toplevel_eval_flex at /home/fredrik/julia15/src/toplevel.c:840
jl_toplevel_eval_flex at /home/fredrik/julia15/src/toplevel.c:790
jl_toplevel_eval_in at /home/fredrik/julia15/src/toplevel.c:883
eval at ./boot.jl:331
_jl_invoke at /home/fredrik/julia15/src/gf.c:2214 [inlined]
jl_apply_generic at /home/fredrik/julia15/src/gf.c:2398
eval_user_input at /home/fredrik/julia15/usr/share/julia/stdlib/v1.5/REPL/src/REPL.jl:134
repl_backend_loop at /home/fredrik/julia15/usr/share/julia/stdlib/v1.5/REPL/src/REPL.jl:195
start_repl_backend at /home/fredrik/julia15/usr/share/julia/stdlib/v1.5/REPL/src/REPL.jl:180
#run_repl#37 at /home/fredrik/julia15/usr/share/julia/stdlib/v1.5/REPL/src/REPL.jl:292
run_repl at /home/fredrik/julia15/usr/share/julia/stdlib/v1.5/REPL/src/REPL.jl:288
_jl_invoke at /home/fredrik/julia15/src/gf.c:2231 [inlined]
jl_apply_generic at /home/fredrik/julia15/src/gf.c:2398
#806 at ./client.jl:399
jfptr_YY.806_46113 at /opt/julia/julia-1.5/lib/julia/sys.so (unknown line)
_jl_invoke at /home/fredrik/julia15/src/gf.c:2214 [inlined]
jl_apply_generic at /home/fredrik/julia15/src/gf.c:2398
jl_apply at /home/fredrik/julia15/src/julia.h:1690 [inlined]
do_apply at /home/fredrik/julia15/src/builtins.c:655
jl_f__apply_latest at /home/fredrik/julia15/src/builtins.c:705
#invokelatest#1 at ./essentials.jl:710 [inlined]
invokelatest at ./essentials.jl:709 [inlined]
run_main_repl at ./client.jl:383
exec_options at ./client.jl:313
_start at ./client.jl:506
jfptr__start_47044 at /opt/julia/julia-1.5/lib/julia/sys.so (unknown line)
_jl_invoke at /home/fredrik/julia15/src/gf.c:2214 [inlined]
jl_apply_generic at /home/fredrik/julia15/src/gf.c:2398
unknown function (ip: 0x401864)
unknown function (ip: 0x401463)
__libc_start_main at /lib/x86_64-linux-gnu/libc.so.6 (unknown line)
unknown function (ip: 0x401508)
Allocations: 782359 (Pool: 781946; Big: 413); GC: 1
Segmentation fault (core dumped)
```

Edit: JuliaInterpreter v0.8.1

### PR DESCRIPTION
**Pull request by goerch**
... by moving `invokelatest` up the call stack.

### CODE REVIEW

**Comment by pfitzseb** [comment]
Why does this work? And how much of a performance impact does it have?

**Comment by goerch** [comment]
> Why does this work? 

I first tried [smaller changes](https://github.com/JuliaDebug/JuliaInterpreter.jl/issues/432#issuecomment-987957947) to repair the  problem, but then some tests failed. I don't have tools available to diagnose the crash in Julia.

> And how much of a performance impact does it have?

Doesn't CI tell us that?

**Comment by goerch** [comment]
Ouch. Any hints how to repair this for Julia 1?

**Comment by pfitzseb** [comment]
Wait for https://github.com/JuliaDebug/JuliaInterpreter.jl/pull/499 to be merged and rebase :)

File: src/interpret.jl [hunk 1]
```diff
@@ -173,12 +173,7 @@ function bypass_builtins(frame, call_expr, pc)
         if isa(tme, Compiled)
             fargs = collect_args(frame, call_expr)
             f = to_function(fargs[1])
-            fmod = parentmodule(f)::Module
-            if fmod === JuliaInterpreter.CompiledCalls || fmod === Core.Compiler
-                return Some{Any}(Base.invokelatest(f, fargs[2:end]...))
-            else
-                return Some{Any}(f(fargs[2:end]...))
-            end
```
Comment (aviatesk) [inline src/interpret.jl:181]: Maybe we want to keep this code around (like with commented out), and leave some comment that says ideally we want to figure out the root problem here and remove the added `invokelatest`.

Comment (goerch) [reply src/interpret.jl]: Done.

**Comment by pfitzseb** [comment]
> I worry this isn't the right fix...it's just kind of a bandaid, and one that will likely hurt performance.

Same, which is why I'd like to understand why this works. 
That said, `invokelatest` performance got much better recently, I think.

**Comment by goerch** [comment]
> You can run benchmarks locally in https://github.com/JuliaDebug/JuliaInterpreter.jl/tree/master/benchmark, although as a caution our benchmarks may not be completely thorough.

For main:

```
Benchmarkresults:
    Package: JuliaInterpreter
    Date: 13 Dec 2021 - 13:34
    Package commit: dirty
    Julia commit: 3bf9d1
    BenchmarkGroup:
        7-element BenchmarkTools.BenchmarkGroup:
          tags: []
          "recursive other 1_000" => Trial(6.007 ms)
          "recursive self 1_000" => Trial(2.848 ms)
          "ccall library" => Trial(82.300 μs)
          "ccall ptr" => Trial(19.900 μs)
          "long function 5_000" => Trial(21.305 ms)
          "throw long 1_000" => Trial(6.804 ms)
          "tight loop 10_000" => Trial(221.751 ms)
```

For branch:

```
Benchmarkresults:
    Package: JuliaInterpreter
    Date: 13 Dec 2021 - 13:40
    Package commit: 7c156d
    Julia commit: 3bf9d1
    BenchmarkGroup:
        7-element BenchmarkTools.BenchmarkGroup:
          tags: []
          "recursive other 1_000" => Trial(5.827 ms)
          "recursive self 1_000" => Trial(3.056 ms)
          "ccall library" => Trial(80.000 μs)
          "ccall ptr" => Trial(18.100 μs)
          "long function 5_000" => Trial(21.651 ms)
          "throw long 1_000" => Trial(6.737 ms)
          "tight loop 10_000" => Trial(233.557 ms)
```

**Comment by goerch** [comment]
> Why does this work? 

At least I now understand that there seems to be some kind of world age confusion. Interesting question I've been pondering is how to explicitly track world age.

**Comment by KristofferC** [comment]
Let's get this in even though we are not sure exactly what is going on. It fixes a problem and doesn't complicate the code much.

**Pull request merged**